internal/provider/resources/resource_api_token.go

```diff
@@ -0,0 +1,566 @@
+		resp.Diagnostics.Append(validateOrganizationApiToken(roles)...)
+	case string(iam.ApiTokenRoleEntityTypeWORKSPACE):
+		resp.Diagnostics.Append(validateWorkspaceApiToken(roles)...)
+	case string(iam.ApiTokenRoleEntityTypeDEPLOYMENT):
+		resp.Diagnostics.Append(validateDeploymentApiToken(roles)...)
+	}
+}
+
+func validateOrganizationApiToken(roles []iam.ApiTokenRole) diag.Diagnostics {
+	for _, role := range roles {
+		if utils.CheckRole(role.Role, string(role.EntityType)) {
+			return nil
+		}
+	}
+	return diag.Diagnostics{
+		diag.NewErrorDiagnostic(
+			"Client Error",
+			"Unable to find the role for the entity type",
+		),
+	}
+}
+
+func validateWorkspaceApiToken(roles []iam.ApiTokenRole) diag.Diagnostics {
+	for _, role := range roles {
+		if role.EntityType == iam.ApiTokenRoleEntityTypeORGANIZATION {
+			return diag.Diagnostics{
+				diag.NewErrorDiagnostic(
+					"Client Error",
+					"API Token of type WORKSPACE cannot have an ORGANIZATION role",
+				),
+			}
+		}
+
+		if utils.CheckRole(role.Role, string(role.EntityType)) {
+			return nil
+		}
+	}
+	return diag.Diagnostics{
+		diag.NewErrorDiagnostic(
+			"Client Error",
+			"Unable to find the role for the entity type",
+		),
+	}
+}
+
+func validateDeploymentApiToken(roles []iam.ApiTokenRole) diag.Diagnostics {
+	for _, role := range roles {
+		if role.EntityType != iam.ApiTokenRoleEntityTypeDEPLOYMENT {
+			return diag.Diagnostics{
+				diag.NewErrorDiagnostic(
+					"Client Error",
+					"API Token of type DEPLOYMENT cannot have an ORGANIZATION or WORKSPACE role",
+				),
+			}
+		}
+
+		if utils.CheckRole(role.Role, string(role.EntityType)) {
+			return nil
+		}
+	}
+	return diag.Diagnostics{
+		diag.NewErrorDiagnostic(
+			"Client Error",
+			"Unable to find the role for the entity type",
+		),
+	}
+}
+
+// RequestApiTokenRoles converts a Terraform set to a list of iam.ApiTokenRole to be used in create and update requests
+func RequestApiTokenRoles(ctx context.Context, apiTokenRolesObjSet types.Set) ([]iam.ApiTokenRole, diag.Diagnostics) {
+	if len(apiTokenRolesObjSet.Elements()) == 0 {
+		return []iam.ApiTokenRole{}, nil
+	}
+
+	var roles []models.ApiTokenRole
+	diags := apiTokenRolesObjSet.ElementsAs(ctx, &roles, false)
+	if diags.HasError() {
+		return nil, diags
+	}
+	apiTokenRoles := lo.Map(roles, func(v models.ApiTokenRole, _ int) iam.ApiTokenRole {
+		var entityType iam.ApiTokenRoleEntityType
+
+		if v.EntityType.ValueString() == string(iam.ApiTokenRoleEntityTypeORGANIZATION) {
+			entityType = iam.ApiTokenRoleEntityTypeORGANIZATION
+		} else if v.EntityType.ValueString() == string(iam.ApiTokenRoleEntityTypeWORKSPACE) {
+			entityType = iam.ApiTokenRoleEntityTypeWORKSPACE
+		} else if v.EntityType.ValueString() == string(iam.ApiTokenRoleEntityTypeDEPLOYMENT) {
+			entityType = iam.ApiTokenRoleEntityTypeDEPLOYMENT
+		}
+
+		return iam.ApiTokenRole{
+			Role:       v.Role.ValueString(),
+			EntityId:   v.EntityId.ValueString(),
+			EntityType: entityType,
+		}
```

Comment:
```suggestion
	apiTokenRoles := lo.Map(roles, func(v models.ApiTokenRole, _ int) iam.ApiTokenRole {
		return iam.ApiTokenRole{
			Role:       v.Role.ValueString(),
			EntityId:   v.EntityId.ValueString(),
			EntityType: iam.ApiTokenRoleEntityType(v.EntityType.ValueString()),
		}
	})
```

Reply:
good point - wasn't aware we could just do it like that